### PR TITLE
Build and push docker image to hub automatically

### PR DESCRIPTION
**Pull request by michaelkaye**
Automate pushing docker images.

Questions - do we want to build any other tags? Is that regex good enough for tags we want to build or do we have a more precise naming scheme. I was hesitant to just look for all tags starting "v" in case we tag something "very_broken_build".

### CODE REVIEW

**Comment by michaelkaye** [comment]
The earlier code mistakenly allowed all branches and only those tags - we're now building on only those tags and no branches.

File: .circleci/config.yml [hunk 1]
```diff
@@ -119,3 +126,10 @@ workflows:
           filters:
             branches:
               ignore: /develop|master/
+      - dockerhubupload:
+          filters:
+            tags:
+              only: /^v[0-9].[0-9]+.[0-9]+(.[0-9]+)?/
+            branches:
+              ignore: /.*/
```
Comment (turt2live) [inline .circleci/config.yml]: Can we ***please*** get a :latest for master?

**Comment by michaelkaye** [comment]
(rebased against develop to pull in hawkowls' changes)

**Comment by michaelkaye** [comment]
This will need two environment variables adding to CircleCI : https://circleci.com/gh/matrix-org/synapse/edit#env-vars

DOCKER_HUB_USERNAME - doesn't need to be matrixdotorg, can be another user who is a contributor to just the synapse project (to minimize access)
DOCKER_HUB_PASSWORD - the user's password to log into docker hub.

We should be aware that someone who can change the .circleci/config.yml and cause a PR or branch to be build which contains, for example: "run: echo $DOCKER_HUB_PASSWORD" will be able to obtain our credentials. This is our responsibility as reviewers to never permit those PRs to be merged to or created within this repository.

We do not pass these secrets to PRs from forks of this repository - so only those with commit or merge access to this repository are able to obtain these secrets - specifically we must never set https://circleci.com/gh/matrix-org/synapse/edit#advanced-settings the "pass secrets to builds of forks of this repository"

**Comment by michaelkaye** [comment]
Let's just leave it on develop and watch it as we make the release candidate.

**Comment by michaelkaye** [comment]
we might need to make a tweak to the rc branch if it doesn't build cleanly, but *shrug*

**Pull request merged**